File: tools/aidkit/pkg/l2/staleness.go
package l2

import (
	"fmt"
	"os/exec"
	"strconv"
	"strings"

	"github.com/dan-strohschein/aidkit/pkg/parser"
)

// StaleClaim represents a claim whose source reference may be outdated.
type StaleClaim struct {
	Entry    string
	Field    string
	Ref      parser.SourceRef
	Reason   string // "file changed", "lines changed", "file deleted"
	ClaimText string
}

// CheckStaleness compares an AID file's @code_version against the current
// git HEAD and reports which [src:] references point to changed code.
func CheckStaleness(aidFile *parser.AidFile, projectRoot string) ([]StaleClaim, error) {
	codeVersion := aidFile.Header.CodeVersion
	if codeVersion == "" {
		return nil, fmt.Errorf("no @code_version in AID file")
	}

	// Extract git hash
	hash := strings.TrimPrefix(codeVersion, "git:")
	if hash == codeVersion {
		return nil, fmt.Errorf("@code_version %q doesn't start with 'git:'", codeVersion)
	}

	// Get current HEAD
	currentHead, err := gitHead(projectRoot)
	if err != nil {
		return nil, fmt.Errorf("get git HEAD: %w", err)
	}

	if strings.HasPrefix(currentHead, hash) || strings.HasPrefix(hash, currentHead) {
		// Same commit — nothing is stale
		return nil, nil
	}

	// Get list of changed files between code_version and HEAD
	changedFiles, err := gitChangedFiles(projectRoot, hash, currentHead)
	if err != nil {
		return nil, fmt.Errorf("git diff: %w", err)
	}

	changedSet := map[string]bool{}
	for _, f := range changedFiles {
		changedSet[f] = true
	}

	// Check each source ref
	var stale []StaleClaim

	checkFields := func(entryName string, fields map[string]parser.Field) {
		for _, field := range fields {
			for _, ref := range field.SourceRefs {
				relPath := ref.File
				if changedSet[relPath] {
					// File was modified — check if the specific lines changed
					linesChanged, _ := gitLinesChanged(projectRoot, hash, currentHead, relPath, ref.StartLine, ref.EndLine)
					if linesChanged {
						stale = append(stale, StaleClaim{
							Entry:     entryName,
							Field:     field.Name,
							Ref:       ref,
							Reason:    "lines changed",
							ClaimText: truncate(field.Value(), 100),
						})
					}
				}
			}
		}
	}

	for _, e := range aidFile.Entries {
		checkFields(e.Name, e.Fields)
	}
	for _, w := range aidFile.Workflows {
		checkFields("workflow:"+w.Name, w.Fields)
	}

	return stale, nil
}

func gitHead(projectRoot string) (string, error) {
	cmd := exec.Command("git", "rev-parse", "--short", "HEAD")
	cmd.Dir = projectRoot
	out, err := cmd.Output()
	if err != nil {
		return "", err
	}
	return strings.TrimSpace(string(out)), nil
}

func gitChangedFiles(projectRoot, fromHash, toHash string) ([]string, error) {
	cmd := exec.Command("git", "diff", "--name-only", fromHash, toHash)
	cmd.Dir = projectRoot
	out, err := cmd.Output()
	if err != nil {
		return nil, err
	}
	lines := strings.Split(strings.TrimSpace(string(out)), "\n")
	var files []string
	for _, l := range lines {
		l = strings.TrimSpace(l)
		if l != "" {
			files = append(files, l)
		}
	}
	return files, nil
}

// gitLinesChanged checks if specific lines in a file were modified between two commits.
func gitLinesChanged(projectRoot, fromHash, toHash, filePath string, startLine, endLine int) (bool, error) {
	// Use git diff with line range
	cmd := exec.Command("git", "diff", fromHash, toHash, "--", filePath)
	cmd.Dir = projectRoot
	out, err := cmd.Output()
	if err != nil {
		return true, nil // Assume changed on error
	}

	diffOutput := string(out)
	if diffOutput == "" {
		return false, nil
	}

	// Parse diff hunks to check if our line range is affected
	for _, line := range strings.Split(diffOutput, "\n") {
		if !strings.HasPrefix(line, "@@") {
			continue
		}
		// Parse @@ -start,count +start,count @@ format
		hunkStart, hunkEnd := parseHunkRange(line)
		if hunkStart <= endLine && hunkEnd >= startLine {
			return true, nil // Hunk overlaps our line range
		}
	}

	return false, nil
}

func parseHunkRange(hunkLine string) (start, end int) {
	// @@ -oldStart,oldCount +newStart,newCount @@
	parts := strings.Split(hunkLine, " ")
	for _, p := range parts {
		if strings.HasPrefix(p, "+") && strings.Contains(p, ",") {
			p = strings.TrimPrefix(p, "+")
			nums := strings.Split(p, ",")
			if len(nums) == 2 {
				s, _ := strconv.Atoi(nums[0])
				c, _ := strconv.Atoi(nums[1])
				return s, s + c - 1
			}
		} else if strings.HasPrefix(p, "+") {
			p = strings.TrimPrefix(p, "+")
			s, _ := strconv.Atoi(p)
			return s, s
		}
	}
	return 0, 0
}

func truncate(s string, maxLen int) string {
	s = strings.ReplaceAll(s, "\n", " ")
	if len(s) > maxLen {
		return s[:maxLen-3] + "..."
	}
	return s
}

File: tools/aidkit-py/src/aidkit/l2/staleness.py
"""Git-based staleness checking for AID source references."""
from __future__ import annotations

import subprocess
from dataclasses import dataclass

from ..model import AidFile, Field, SourceRef


@dataclass
class StaleClaim:
    """A claim whose source reference may be outdated."""

    entry: str
    field: str
    ref: SourceRef
    reason: str  # "file changed", "lines changed", "file deleted"
    claim_text: str


def check_staleness(aid_file: AidFile, project_root: str) -> list[StaleClaim]:
    """Compare an AID file's @code_version against the current git HEAD.

    Reports which [src:] references point to changed code.

    Raises ValueError if @code_version is missing or malformed.
    """
    code_version = aid_file.header.code_version
    if not code_version:
        raise ValueError("no @code_version in AID file")

    if not code_version.startswith("git:"):
        raise ValueError(f"@code_version {code_version!r} doesn't start with 'git:'")

    commit_hash = code_version[4:]

    current_head = _git_head(project_root)
    if current_head.startswith(commit_hash) or commit_hash.startswith(current_head):
        return []

    changed_files = _git_changed_files(project_root, commit_hash, current_head)
    changed_set = set(changed_files)

    stale: list[StaleClaim] = []

    def check_fields(entry_name: str, fields: dict[str, Field]) -> None:
        for field in fields.values():
            for ref in field.source_refs:
                if ref.file in changed_set:
                    lines_changed = _git_lines_changed(
                        project_root, commit_hash, current_head,
                        ref.file, ref.start_line, ref.end_line,
                    )
                    if lines_changed:
                        stale.append(StaleClaim(
                            entry=entry_name,
                            field=field.name,
                            ref=ref,
                            reason="lines changed",
                            claim_text=_truncate(field.value(), 100),
                        ))

    for e in aid_file.entries:
        check_fields(e.name, e.fields)
    for w in aid_file.workflows:
        check_fields(f"workflow:{w.name}", w.fields)

    return stale


def _git_head(project_root: str) -> str:
    result = subprocess.run(
        ["git", "rev-parse", "--short", "HEAD"],
        cwd=project_root, capture_output=True, text=True, check=True,
    )
    return result.stdout.strip()


def _git_changed_files(project_root: str, from_hash: str, to_hash: str) -> list[str]:
    result = subprocess.run(
        ["git", "diff", "--name-only", from_hash, to_hash],
        cwd=project_root, capture_output=True, text=True, check=True,
    )
    return [l.strip() for l in result.stdout.strip().split("\n") if l.strip()]


def _git_lines_changed(
    project_root: str,
    from_hash: str,
    to_hash: str,
    file_path: str,
    start_line: int,
    end_line: int,
) -> bool:
    """Check if specific lines in a file were modified between two commits."""
    try:
        result = subprocess.run(
            ["git", "diff", from_hash, to_hash, "--", file_path],
            cwd=project_root, capture_output=True, text=True, check=True,
        )
    except subprocess.CalledProcessError:
        return True  # Assume changed on error

    if not result.stdout:
        return False

    for line in result.stdout.split("\n"):
        if not line.startswith("@@"):
            continue
        hunk_start, hunk_end = _parse_hunk_range(line)
        if hunk_start <= end_line and hunk_end >= start_line:
            return True

    return False


def _parse_hunk_range(hunk_line: str) -> tuple[int, int]:
    """Parse @@ -oldStart,oldCount +newStart,newCount @@ format."""
    for part in hunk_line.split():
        if part.startswith("+") and "," in part:
            part = part.lstrip("+")
            nums = part.split(",")
            if len(nums) == 2:
                s = int(nums[0])
                c = int(nums[1])
                return s, s + c - 1
        elif part.startswith("+"):
            part = part.lstrip("+")
            try:
                s = int(part)
                return s, s
            except ValueError:
                continue
    return 0, 0


def _truncate(s: str, max_len: int) -> str:
    s = s.replace("\n", " ")
    if len(s) > max_len:
        return s[: max_len - 3] + "..."
    return s
